File: backend_python/app/services/service_service.py
"""
Service listing service
"""
from typing import Optional, List, Any
from datetime import datetime
from app.models.service_listing import ServiceListing, ServiceStatus
from app.models.user import User, Role, UserRole
from app.models.service_provider import ServiceProvider, ProviderStatus
from app.models.verification import VerificationAssignment, VerificationStatus
from app.models.nodal_officer import (
    NodalOfficerAssignment,
    NodalOfficerRecommendation,
    NodalOfficerStatus,
    RecommendationStatus,
)
from app.models.manager import ManagerAssignment
from app.models.service_category import ServiceCategory
from app.middleware.error_handler import HttpError
from app.utils.jwt import sign_token
from app.config.env import settings
from app.config.logger import setup_logger

logger = setup_logger()


class ServiceService:
    """Service listing operations"""
    
    @staticmethod
    async def create_service(
        owner_id: str,
        data: dict,
    ) -> ServiceListing:
        """Create a new service listing"""
        service = ServiceListing(
            ownerId=owner_id,
            title=data["title"],
            description=data["description"],
            location=data["location"],
            address=data.get("address"),
            pincode=data.get("pincode"),
            eventTypes=data.get("eventTypes"),
            propertyType=data.get("propertyType"),
            capacity=data.get("capacity"),
            amenities=data.get("amenities"),
            photos=data.get("photos"),
            price=data["price"],
            status=ServiceStatus.PENDING,
        )
        await service.insert()
        return service
    
    @staticmethod
    async def list_user_services(owner_id: str) -> List[ServiceListing]:
        """List services owned by a user"""
        services = await ServiceListing.find(
            ServiceListing.owner_id == owner_id
        ).to_list()
        return services
    
    @staticmethod
    async def list_nodal_officer_services(officer_id: str) -> List[ServiceListing]:
        """List services assigned to a nodal officer"""
        services = await ServiceListing.find(
            ServiceListing.assigned_nodal_officer_id == officer_id
        ).to_list()
        return services
    
    @staticmethod
    async def list_pending() -> List[ServiceListing]:
        """List pending services"""
        # Use dictionary query for Beanie
        services = await ServiceListing.find(
            {"status": "PENDING"}
        ).to_list()
        return services
    
    @staticmethod
    async def list_published() -> List[ServiceListing]:
        """List published services"""
        # Use dictionary query for Beanie
        services = await ServiceListing.find(
            {"status": "PUBLISHED"}
        ).to_list()
        return services
    
    @staticmethod
    async def list_managed_services(manager_id: str) -> List[dict]:
        """List services managed by a manager"""
        services = await ServiceListing.find(
            {"assignedManagerId": manager_id}
        ).to_list()
        
        result = []
        for service in services:
            owner = await User.get(service.owner_id) if service.owner_id else None
            assigned_manager = await User.get(service.assigned_manager_id) if service.assigned_manager_id else None
            assigned_nodal_officer = await User.get(service.assigned_nodal_officer_id) if service.assigned_nodal_officer_id else None
            
            result.append({
                **service.dict(),
                "owner": {
                    "id": str(owner.id) if owner else None,
                    "fullName": owner.full_name if owner else None,
                    "email": owner.email if owner else None,
                } if owner else None,
                "assignedManager": assigned_manager.dict() if assigned_manager else None,
                "assignedNodalOfficer": assigned_nodal_officer.dict() if assigned_nodal_officer else None,
            })
        
        return result
    
    @staticmethod
    async def assign_nodal_officer(
        manager_id: str,
        service_id: str,
        nodal_officer_id: str,
    ) -> dict:
        """Assign a nodal officer to a service"""
        service = await ServiceListing.get(service_id)
        if not service:
            raise HttpError(404, "Service not found")
        
        # Verify nodal officer exists and is active
        officer_assignment = await NodalOfficerAssignment.find_one(
            {"officerId": nodal_officer_id, "status": "ACTIVE"}
        )
        
        if not officer_assignment:
            raise HttpError(404, "Nodal officer not found or not active")
        
        # Update service
        service.status = ServiceStatus.ASSIGNED
        service.assigned_nodal_officer_id = nodal_officer_id
        await service.save()
        
        # Send notification
        await ServiceService._send_nodal_officer_notification(
            officer_assignment.officer_id,
            service,
        )
        
        # Fetch updated service with relations
        assigned_nodal_officer = await User.get(nodal_officer_id)
        
        return {
            "service": service.dict(),
            "nodalOfficer": assigned_nodal_officer.dict() if assigned_nodal_officer else None,
        }
    
    @staticmethod
    async def _send_nodal_officer_notification(
        officer_id: str,
        service: ServiceListing,
    ):
        """Send notification to nodal officer"""
        officer = await User.get(officer_id)
        if not officer:
            return
        
        # Generate activation token
        activation_token = sign_token(
            {"userId": str(officer.id), "type": "password_setup"},
            expires_in="7d",
        )
        
        frontend_url = settings.frontend_url or "http://localhost"
        activation_link = f"{frontend_url}/auth/setup-password?token={activation_token}"
        
        logger.info("Nodal Officer Assignment Notification", {
            "officerEmail": officer.email,
            "officerPhone": officer.phone,
            "serviceTitle": service.title,
            "serviceLocation": service.location,
            "activationLink": activation_link,
        })
        
        # TODO: Send email notification
        # from app.utils.email import send_nodal_officer_assignment_email
        # await send_nodal_officer_assignment_email(...)
    
    @staticmethod
    async def assign_agents(
        manager_id: str,
        service_id: str,
        agent_ids: List[str],
        nodal_officer_id: Optional[str] = None,
    ) -> dict:
        """Assign verification agents to a service"""
        service = await ServiceListing.get(service_id)
        if not service:
            raise HttpError(404, "Service not found")
        
        # Find available officers
        available_officers = await ServiceService.find_nodal_officers_by_location(
            service.location,
            service.pincode,
        )
        
        final_nodal_officer_id = nodal_officer_id
        
        # Auto-assign if no officer provided
        if not final_nodal_officer_id and available_officers:
            final_nodal_officer_id = available_officers[0]["officerId"]
        
        # Verify officer is in area
        if final_nodal_officer_id:
            officer_exists = any(
                o["officerId"] == final_nodal_officer_id for o in available_officers
            )
            if not officer_exists:
                raise HttpError(
                    400,
                    "Selected nodal officer is not assigned to this service area",
                )
        
        # Update service status
        service.status = ServiceStatus.ASSIGNED
        await service.save()
        
        # Create verification assignments
        for provider_id in agent_ids:
            assignment = VerificationAssignment(
                serviceId=service_id,
                providerId=provider_id,
                assignedById=manager_id,
                status=VerificationStatus.PENDING,
            )
            await assignment.insert()
        
        return {
            "service": service.dict(),
            "availableNodalOfficers": available_officers,
            "assignedNodalOfficerId": final_nodal_officer_id,
            "needsRecommendation": len(available_officers) == 0,
        }
    
    @staticmethod
    async def find_nodal_officers_by_location(
        location: str,
        pincode: Optional[str] = None,
    ) -> List[dict]:
        """Find nodal officers by service location"""
        where_clause = {"status": NodalOfficerStatus.ACTIVE}
        if pincode:
            where_clause["pincode"] = pincode
        
        officers = await NodalOfficerAssignment.find(where_clause).to_list()
        
        # Filter by location case-insensitively
        filtered = [
            o for o in officers
            if location.lower() in o.location.lower()
        ]
        
        result = []
        for officer in filtered:
            officer_user = await User.get(officer.officer_id)
            result.append({
                "officerId": str(officer.officer_id),
                "officer": officer_user.dict() if officer_user else None,
                "area": officer.area,
                "location": officer.location,
            })
        
        return result
    
    @staticmethod
    async def verify_service(
        assignment_id: str,
        decision: VerificationStatus,
        notes: Optional[str] = None,
    ):
        """Verify a service assignment"""
        assignment = await VerificationAssignment.get(assignment_id)
        if not assignment:
            raise HttpError(404, "Assignment not found")
        
        assignment.status = decision
        if notes:
            assignment.notes = notes
        await assignment.save()
        
        if decision == VerificationStatus.APPROVED:
            # Count approvals
            approvals = await VerificationAssignment.find(
                {"serviceId": assignment.service_id, "status": "APPROVED"}
            ).count()
            
            if approvals >= 3:
                service = await ServiceListing.get(assignment.service_id)
                if service:
                    service.status = ServiceStatus.PUBLISHED
                    service.published_at = datetime.utcnow()
                    await service.save()
    
    @staticmethod
    async def list_assignments(provider_id: str) -> List[VerificationAssignment]:
        """List verification assignments for a provider"""
        assignments = await VerificationAssignment.find(
            {"providerId": provider_id}
        ).to_list()
        return assignments
    
    @staticmethod
    async def get_service_verification_assignments(service_id: str) -> List[dict]:
        """Get verification assignments for a service (for admins/managers to view progress)"""
        assignments = await VerificationAssignment.find(
            {"serviceId": service_id}
        ).to_list()
        
        result = []
        for assignment in assignments:
            # Get provider user info
            provider_user = await User.get(assignment.provider_id) if assignment.provider_id else None
            assigned_by_user = await User.get(assignment.assigned_by_id) if assignment.assigned_by_id else None
            
            assignment_dict = assignment.dict()
            assignment_dict["provider"] = {
                "id": str(provider_user.id),
                "email": provider_user.email,
                "fullName": provider_user.full_name,
                "phone": provider_user.phone,
            } if provider_user else None
            
            assignment_dict["assignedBy"] = {
                "id": str(assigned_by_user.id),
                "email": assigned_by_user.email,
                "fullName": assigned_by_user.full_name,
            } if assigned_by_user else None
            
            result.append(assignment_dict)
        
        return result
    
    @staticmethod
    async def onboard_provider(
        user_id: str,
        data: dict,
    ) -> ServiceProvider:
        """Onboard a service provider"""
        # Create/update role
        role = await Role.find_one(
            {"userId": user_id, "role": "PROVIDER"}
        )
        if not role:
            role = Role(userId=user_id, role=UserRole.PROVIDER)
            await role.insert()
        
        # Create/update provider profile
        provider = await ServiceProvider.find_one({"userId": user_id})
        if provider:
            provider.kyc_document_url = data.get("kycDocumentUrl")
            provider.pricing_details = data.get("pricingDetails")
            provider.photos = data.get("photoUrls", [])
            provider.status = ProviderStatus.PENDING
            await provider.save()
        else:
            provider = ServiceProvider(
                userId=user_id,
                kycDocumentUrl=data.get("kycDocumentUrl"),
                pricingDetails=data.get("pricingDetails"),
                photos=data.get("photoUrls", []),
                status=ProviderStatus.PENDING,
            )
            await provider.insert()
        
        return provider
    
    @staticmethod
    async def list_providers() -> List[dict]:
        """List approved providers"""
        try:
            providers = await ServiceProvider.find(
                {"status": "APPROVED"}
            ).to_list()
            
            result = []
            for provider in providers:
                try:
                    # Use find_one instead of get to avoid _id issues
                    user = None
                    if provider.user_id:
                        user = await User.find_one({"_id": provider.user_id})
                        if not user:
                            # Try as string ID
                            try:
                                from bson import ObjectId
                                user = await User.find_one({"_id": ObjectId(provider.user_id)})
                            except:
                                pass
                    
                    # Convert to dict, handling ObjectId serialization
                    provider_dict = {
                        "id": str(provider.id),
                        "userId": str(provider.user_id) if provider.user_id else None,
                        "kycDocumentUrl": provider.kyc_document_url,
                        "pricingDetails": provider.pricing_details,
                        "photos": provider.photos or [],
                        "status": provider.status.value if hasattr(provider.status, "value") else str(provider.status),
                    }
                    
                    user_dict = None
                    if user:
                        user_dict = {
                            "id": str(user.id),
                            "email": user.email,
                            "fullName": user.full_name,
                            "phone": user.phone,
                        }
                    
                    result.append({
                        **provider_dict,
                        "user": user_dict,
                    })
                except Exception as e:
                    logger.error(f"Error processing provider {provider.id}: {e}", exc_info=True)
                    continue
            
            return result
        except Exception as e:
            logger.error(f"Error in list_providers: {e}", exc_info=True)
            raise
    
    @staticmethod
    async def list_nodal_officers_by_area(
        manager_id: str,
        filters: Optional[dict] = None,
    ) -> List[dict]:
        """List nodal officers by area (for managers)"""
        manager_assignment = await ManagerAssignment.find_one({
            "managerId": manager_id
        })
        
        if not manager_assignment:
            raise HttpError(
                404,
                "Manager assignment not found. Please contact admin to assign you to a region.",
            )
        
        if manager_assignment.status != ManagerAssignmentStatus.ACTIVE:
            raise HttpError(403, "Manager assignment is not active")
        
        # Build filter
        where_clause = {"status": NodalOfficerStatus.ACTIVE}
        if manager_assignment.pincode:
            where_clause["pincode"] = manager_assignment.pincode
        
        officers = await NodalOfficerAssignment.find(where_clause).to_list()
        
        # Filter by location
        location_filter = filters.get("location") if filters else manager_assignment.location
        if location_filter:
            officers = [
                o for o in officers
                if location_filter.lower() in o.location.lower()
            ]
        
        # Additional filters
        if filters:
            if filters.get("area"):
                officers = [
                    o for o in officers
                    if filters["area"].lower() in o.area.lower()
                ]
            if filters.get("pincode"):
                officers = [o for o in officers if o.pincode == filters["pincode"]]
        
        result = []
        for officer in officers:
            officer_user = await User.get(officer.officer_id)
            result.append({
                **officer.dict(),
                "officer": officer_user.dict() if officer_user else None,
            })
        
        return result
    
    @staticmethod
    async def recommend_nodal_officer(
        manager_id: str,
        data: dict,
    ) -> NodalOfficerRecommendation:
        """Recommend a nodal officer"""
        if not data.get("officerId") and not (data.get("newUserEmail") and data.get("newUserFullName")):
            raise HttpError(
                400,
                "Either officerId or new user details (email, fullName) must be provided",
            )
        
        # Validate existing officer
        if data.get("officerId"):
            officer = await User.get(data["officerId"])
            if not officer:
                raise HttpError(404, "Officer not found")
            
            roles = await Role.find({"userId": str(officer.id)}).to_list()
            has_role = any(role.role == UserRole.NODAL_OFFICER for role in roles)
            if not has_role:
                raise HttpError(400, "User does not have NODAL_OFFICER role")
        
        # Check for existing recommendation
        existing = await NodalOfficerRecommendation.find_one(
            {"managerId": manager_id, "area": data["area"], "location": data["location"], "status": "PENDING"}
        )
        
        if existing:
            raise HttpError(409, "Recommendation already pending for this officer and area")
        
        # Create recommendation
        recommendation = NodalOfficerRecommendation(
            managerId=manager_id,
            officerId=data.get("officerId"),
            newUserEmail=data.get("newUserEmail"),
            newUserFullName=data.get("newUserFullName"),
            newUserPhone=data.get("newUserPhone"),
            area=data["area"],
            location=data["location"],
            address=data.get("address"),
            pincode=data.get("pincode"),
            region=data.get("region"),
            reason=data.get("reason"),
            status=RecommendationStatus.PENDING,
        )
        await recommendation.insert()
        
        return recommendation
    
    @staticmethod
    async def list_manager_recommendations(manager_id: str) -> List[dict]:
        """List recommendations by a manager"""
        recommendations = await NodalOfficerRecommendation.find(
            {"managerId": manager_id}
        ).to_list()
        
        result = []
        for rec in recommendations:
            officer = await User.get(rec.officer_id) if rec.officer_id else None
            result.append({
                **rec.dict(),
                "officer": officer.dict() if officer else None,
            })
        
        return result
    
    @staticmethod
    async def get_manager_assignment(manager_id: str) -> Optional[dict]:
        """Get manager assignment"""
        assignment = await ManagerAssignment.find_one({
            "managerId": manager_id
        })
        return assignment.dict() if assignment else None
    
    @staticmethod
    async def list_navbar_categories() -> List[dict]:
        """List service categories for navbar"""
        categories = await ServiceCategory.find(
            {"showInNavbar": True, "isActive": True}
        ).to_list()
        
        # Sort by display_order
        categories.sort(key=lambda c: c.display_order)
        
        return [
            {
                "id": str(c.id),
                "name": c.name,
                "description": c.description,
                "displayOrder": c.display_order,
            }
            for c in categories
        ]
    
    @staticmethod
    async def get_available_nodal_officers_for_service(service_id: str) -> dict:
        """Get available nodal officers for a service"""
        service = await ServiceListing.get(service_id)
        if not service:
            raise HttpError(404, "Service not found")
        
        officers = await ServiceService.find_nodal_officers_by_location(
            service.location,
            service.pincode,
        )
        
        return {
            "availableOfficers": officers,
            "needsRecommendation": len(officers) == 0,
            "serviceLocation": service.location,
            "servicePincode": service.pincode,
        }

File: backend/src/modules/service/service.service.ts
import type { PrismaClient, ServiceStatus, VerificationStatus } from '@prisma/client';

import { HttpError } from '../../middleware/errorHandler.js';

export async function createService(
  prisma: PrismaClient,
  ownerId: string,
  data: {
    title: string;
    description: string;
    location: string;
    address?: string;
    pincode?: string;
    eventTypes?: string;
    propertyType?: string;
    capacity?: number;
    amenities?: string;
    photos?: string[];
    price: number;
  },
) {
  return prisma.serviceListing.create({
    data: {
      ownerId,
      title: data.title,
      description: data.description,
      location: data.location,
      address: data.address,
      pincode: data.pincode,
      eventTypes: data.eventTypes,
      propertyType: data.propertyType,
      capacity: data.capacity,
      amenities: data.amenities,
      photos: data.photos ? (data.photos as any) : null,
      price: data.price,
    },
  });
}

export function listUserServices(prisma: PrismaClient, ownerId: string) {
  return prisma.serviceListing.findMany({ 
    where: { ownerId },
    include: { owner: { include: { roles: true } } },
    orderBy: { submittedAt: 'desc' },
  });
}

// Fetch services assigned to a nodal officer
export function listNodalOfficerServices(prisma: PrismaClient, officerId: string) {
  return prisma.serviceListing.findMany({
    where: { assignedNodalOfficerId: officerId } as any,
    include: {
      owner: { include: { roles: true } },
      assignedManager: { include: { roles: true } },
      assignedNodalOfficer: { include: { roles: true } },
    } as any,
    orderBy: { submittedAt: 'desc' },
  });
}

export function listPending(prisma: PrismaClient) {
  return prisma.serviceListing.findMany({ where: { status: 'PENDING' } });
}

export function listPublished(prisma: PrismaClient) {
  return prisma.serviceListing.findMany({ where: { status: 'PUBLISHED' } });
}

export function listManagedServices(prisma: PrismaClient, managerId: string) {
  return prisma.serviceListing.findMany({
    where: { assignedManagerId: managerId },
    include: { 
      owner: { select: { id: true, fullName: true, email: true } },
      assignedManager: { include: { roles: true } },
      assignedNodalOfficer: { include: { roles: true } },
    } as any,
  });
}

// New function to assign nodal officer directly (without providers)
export async function assignNodalOfficer(
  prisma: PrismaClient,
  managerId: string,
  serviceId: string,
  nodalOfficerId: string,
) {
  const service = await prisma.serviceListing.findUnique({ where: { id: serviceId } });
  if (!service) {
    throw new HttpError(404, 'Service not found');
  }

  // Verify nodal officer exists and is active
  const officerAssignment = await prisma.nodalOfficerAssignment.findFirst({
    where: {
      officerId: nodalOfficerId,
      status: 'ACTIVE',
    },
    include: {
      officer: {
        include: { roles: true },
      },
    },
  });

  if (!officerAssignment) {
    throw new HttpError(404, 'Nodal officer not found or not active');
  }

  // Update service with nodal officer assignment
  await prisma.serviceListing.update({
    where: { id: serviceId },
    data: {
      status: 'ASSIGNED',
      assignedNodalOfficerId: nodalOfficerId,
    } as any,
  });

  // Send notification to nodal officer (email/SMS)
  await sendNodalOfficerAssignmentNotification(
    prisma,
    officerAssignment.officer,
    service,
  );

  return {
    service: await prisma.serviceListing.findUnique({
      where: { id: serviceId },
      include: {
        assignedManager: true,
        assignedNodalOfficer: {
          include: { roles: true },
        },
      } as any,
    }),
    nodalOfficer: officerAssignment.officer,
  };
}

// Helper function to send notification to nodal officer
async function sendNodalOfficerAssignmentNotification(
  prisma: PrismaClient,
  officer: { id: string; email: string; fullName: string; phone?: string | null },
  service: { id: string; title: string; location: string; address?: string | null },
) {
  // Generate activation token for password setup
  const jwt = await import('jsonwebtoken');
  const { env } = await import('../../config/env.js');
  const activationToken = jwt.sign(
    { userId: officer.id, type: 'password_setup' } as object,
    env.jwtSecret,
    { expiresIn: '7d' },
  );

  const { logger } = await import('../../config/logger.js');
  const { sendNodalOfficerAssignmentEmail } = await import('../../utils/email.js');

  const frontendUrl = env.frontendUrl || 'http://localhost';
  const activationLink = `${frontendUrl}/auth/setup-password?token=${activationToken}`;

  // Log notification details
  logger.info('Nodal Officer Assignment Notification', {
    officerEmail: officer.email,
    officerPhone: officer.phone,
    serviceTitle: service.title,
    serviceLocation: service.location,
    activationLink,
  });

  // Send email notification
  await sendNodalOfficerAssignmentEmail(
    officer.email,
    officer.fullName,
    service.title,
    service.location,
    activationLink,
  );

  // TODO: Integrate with SMS service (e.g., Twilio, AWS SNS) if phone is provided
  // if (officer.phone) {
  //   await sendSMS(officer.phone, `You've been assigned as Nodal Officer. Setup: ${activationLink}`);
  // }
}

export async function assignAgents(
  prisma: PrismaClient,
  managerId: string,
  serviceId: string,
  agentIds: string[],
  nodalOfficerId?: string,
) {
  const service = await prisma.serviceListing.findUnique({ where: { id: serviceId } });
  if (!service) {
    throw new HttpError(404, 'Service not found');
  }

  // Find nodal officers in the service's area
  const availableOfficers = await findNodalOfficersByServiceLocation(
    prisma,
    service.location,
    service.pincode,
  );

  let finalNodalOfficerId = nodalOfficerId;

  // If no officer provided but officers exist in area, auto-assign the first one
  if (!finalNodalOfficerId && availableOfficers.length > 0) {
    finalNodalOfficerId = availableOfficers[0].officerId;
  }

  // If nodal officer is provided, verify it exists and is in the area
  if (finalNodalOfficerId) {
    const officerExists = availableOfficers.some(
      (o) => o.officerId === finalNodalOfficerId,
    );
    if (!officerExists) {
      throw new HttpError(
        400,
        'Selected nodal officer is not assigned to this service area. Please recommend them first or select an officer from the available list.',
      );
    }
  }

  await prisma.$transaction([
    prisma.serviceListing.update({
      where: { id: serviceId },
      data: { status: 'ASSIGNED' },
    }),
    prisma.verificationAssignment.createMany({
      data: agentIds.map((providerId) => ({
        serviceId,
        providerId,
        assignedById: managerId,
      })),
    }),
  ]);

  return {
    service: await prisma.serviceListing.findUnique({
      where: { id: serviceId },
      include: { assignedManager: true },
    }),
    availableNodalOfficers: availableOfficers,
    assignedNodalOfficerId: finalNodalOfficerId || null,
    needsRecommendation: availableOfficers.length === 0,
  };
}

// Helper function to find nodal officers by service location
export async function findNodalOfficersByServiceLocation(
  prisma: PrismaClient,
  location: string,
  pincode?: string | null,
) {
  // For MongoDB, fetch all active officers and filter by location
  const whereClause: any = {
    status: 'ACTIVE',
  };

  if (pincode) {
    whereClause.pincode = pincode;
  }

  const officers = await prisma.nodalOfficerAssignment.findMany({
    where: whereClause,
    include: {
      officer: {
        include: { roles: true },
      },
    },
    orderBy: { assignedAt: 'desc' },
  });

  // Filter by location case-insensitively (MongoDB doesn't support mode: 'insensitive' directly)
  return officers.filter((o) =>
    o.location.toLowerCase().includes(location.toLowerCase()),
  );
}

export async function verifyService(
  prisma: PrismaClient,
  assignmentId: string,
  decision: VerificationStatus,
  notes?: string,
) {
  const assignment = await prisma.verificationAssignment.findUnique({
    where: { id: assignmentId },
  });

  if (!assignment) {
    throw new HttpError(404, 'Assignment not found');
  }

  await prisma.verificationAssignment.update({
    where: { id: assignmentId },
    data: { status: decision, notes },
  });

  if (decision === 'APPROVED') {
    const approvals = await prisma.verificationAssignment.count({
      where: { serviceId: assignment.serviceId, status: 'APPROVED' },
    });

    if (approvals >= 3) {
      await prisma.serviceListing.update({
        where: { id: assignment.serviceId },
        data: { status: 'PUBLISHED', publishedAt: new Date() },
      });
    }
  }
}

export function listAssignments(prisma: PrismaClient, providerId: string) {
  return prisma.verificationAssignment.findMany({
    where: { providerId },
    include: { service: true },
  });
}

export async function onboardProvider(
  prisma: PrismaClient,
  userId: string,
  data: { kycDocumentUrl: string; pricingDetails: string; photoUrls?: string[] },
) {
  return prisma.$transaction(async (tx) => {
    await tx.role.upsert({
      where: { userId_role: { userId, role: 'PROVIDER' } },
      update: {},
      create: { userId, role: 'PROVIDER' },
    });

    return tx.serviceProvider.upsert({
      where: { userId },
      update: {
        kycDocumentUrl: data.kycDocumentUrl,
        pricingDetails: data.pricingDetails,
        photos: data.photoUrls ?? [],
        status: 'PENDING',
      },
      create: {
        userId,
        kycDocumentUrl: data.kycDocumentUrl,
        pricingDetails: data.pricingDetails,
        photos: data.photoUrls ?? [],
      },
    });
  });
}

export function listProviders(prisma: PrismaClient) {
  return prisma.serviceProvider.findMany({
    include: { user: { include: { roles: true } } },
    where: { status: 'APPROVED' },
    orderBy: { createdAt: 'desc' },
  });
}

// Nodal Officer Functions for Managers
export async function listNodalOfficersByArea(
  prisma: PrismaClient,
  managerId: string,
  filters?: { location?: string; area?: string; pincode?: string },
) {
  // Get manager's assignment to filter officers by manager's location
  const managerAssignment = await prisma.managerAssignment.findUnique({
    where: { managerId },
    include: { manager: { include: { roles: true } } },
  });

  if (!managerAssignment) {
    throw new HttpError(404, 'Manager assignment not found. Please contact admin to assign you to a region.');
  }

  if (managerAssignment.status !== 'ACTIVE') {
    throw new HttpError(403, 'Manager assignment is not active');
  }

  // Build filter based on manager's location and optional filters
  // For MongoDB, we'll fetch all and filter in memory for case-insensitive matching
  const whereClause: any = {
    status: 'ACTIVE',
  };

  // If manager has a pincode, filter by it
  if (managerAssignment.pincode) {
    whereClause.pincode = managerAssignment.pincode;
  }

  // Fetch all active officers first
  let officers = await prisma.nodalOfficerAssignment.findMany({
    where: whereClause,
    include: {
      officer: {
        include: { roles: true },
      },
    },
    orderBy: { assignedAt: 'desc' },
  });

  // Filter by location case-insensitively
  const locationFilter = filters?.location || managerAssignment.location;
  if (locationFilter) {
    officers = officers.filter((o) =>
      o.location.toLowerCase().includes(locationFilter.toLowerCase()),
    );
  }

  // Filter by region if provided
  const regionFilter = managerAssignment.region;
  if (regionFilter) {
    officers = officers.filter((o) =>
      o.region?.toLowerCase().includes(regionFilter.toLowerCase()),
    );
  }

  // Filter by area if provided
  if (filters?.area) {
    officers = officers.filter((o) =>
      o.area.toLowerCase().includes(filters.area!.toLowerCase()),
    );
  }

  // Filter by pincode if provided
  if (filters?.pincode) {
    officers = officers.filter((o) => o.pincode === filters.pincode);
  }

  return officers;
}

export async function recommendNodalOfficer(
  prisma: PrismaClient,
  managerId: string,
  data: {
    officerId?: string; // Optional - for existing officers
    newUserEmail?: string; // For new user recommendations
    newUserFullName?: string;
    newUserPhone?: string;
    area: string;
    location: string;
    address?: string;
    pincode?: string;
    region?: string;
    reason?: string;
  },
) {
  // Either officerId OR new user details must be provided
  if (!data.officerId && (!data.newUserEmail || !data.newUserFullName)) {
    throw new HttpError(400, 'Either officerId or new user details (email, fullName) must be provided');
  }

  // If recommending existing officer, verify they exist and have NODAL_OFFICER role
  if (data.officerId) {
    const officer = await prisma.user.findUnique({
      where: { id: data.officerId },
      include: { roles: true },
    });

    if (!officer) {
      throw new HttpError(404, 'Officer not found');
    }

    const hasOfficerRole = officer.roles.some((role) => role.role === 'NODAL_OFFICER');
    if (!hasOfficerRole) {
      throw new HttpError(400, 'User does not have NODAL_OFFICER role');
    }

    // Check if recommendation already exists for existing officer
    const existing = await prisma.nodalOfficerRecommendation.findFirst({
      where: {
        managerId,
        officerId: data.officerId,
        area: data.area,
        location: data.location,
        status: 'PENDING',
      },
    });

    if (existing) {
      throw new HttpError(409, 'Recommendation already pending for this officer and area');
    }
  } else {
    // For new user recommendations, check if email already exists
    const existingUser = await prisma.user.findUnique({
      where: { email: data.newUserEmail! },
    });

    if (existingUser) {
      throw new HttpError(409, 'User with this email already exists');
    }

    // Check if recommendation already exists for this new user email and area
    // Note: MongoDB/Prisma doesn't support filtering by newUserEmail directly in where clause
    // We'll fetch all pending recommendations and filter in memory
    const allPending = await prisma.nodalOfficerRecommendation.findMany({
      where: {
        managerId,
        area: data.area,
        location: data.location,
        status: 'PENDING',
      },
    });
    
    const existing = allPending.find(
      (rec: any) => rec.newUserEmail === data.newUserEmail
    );

    if (existing) {
      throw new HttpError(409, 'Recommendation already pending for this user and area');
    }
  }

  const createData: any = {
    managerId,
    area: data.area,
    location: data.location,
    reason: data.reason,
  };

  if (data.officerId) {
    createData.officerId = data.officerId;
  } else {
    createData.newUserEmail = data.newUserEmail;
    createData.newUserFullName = data.newUserFullName;
    if (data.newUserPhone) {
      createData.newUserPhone = data.newUserPhone;
    }
  }

  if (data.address) createData.address = data.address;
  if (data.pincode) createData.pincode = data.pincode;
  if (data.region) createData.region = data.region;

  return prisma.nodalOfficerRecommendation.create({
    data: createData,
    include: {
      officer: { include: { roles: true } },
      manager: { include: { roles: true } },
    },
  });
}

export function listManagerRecommendations(
  prisma: PrismaClient,
  managerId: string,
) {
  return prisma.nodalOfficerRecommendation.findMany({
    where: { managerId },
    include: {
      officer: { include: { roles: true } },
    },
    orderBy: { createdAt: 'desc' },
  });
}

export function getManagerAssignment(
  prisma: PrismaClient,
  managerId: string,
) {
  return prisma.managerAssignment.findUnique({
    where: { managerId },
    include: {
      manager: { include: { roles: true } },
    },
  });
}

// Service Category Functions - Public endpoint for navbar
export function listNavbarServiceCategories(prisma: PrismaClient) {
  return prisma.serviceCategory.findMany({
    where: {
      showInNavbar: true,
      isActive: true,
    },
    orderBy: [{ displayOrder: 'asc' }, { name: 'asc' }],
    select: {
      id: true,
      name: true,
      description: true,
      displayOrder: true,
    },
  });
}

